### DISCUSSION TITLE
Yapay Zeka için Kaynak Önerisi

### DISCUSSION THREAD

**Discussion by BunyaminAkdeniz** [19/classroom#71] [Anonim]
Merhabalar Hocam. Yapay zeka öğrenmek için python öğrendim fakat ai kısmı için nereden başlayacağıma karar veremedim. Çok fazla kütüphane ve farklı farklı eğitimler var. Bu konuyla alakalı bir eğitim, doküman ya da yol haritası olacak bir tavsiyeleriniz var mı? Teşekkürler.

**Comment by roktas** [comment]
Merhaba,

AI konusunda o kadar ve o kadar çok kaynak var ki kürasyonu zor. Github'ta bir "awesome" geleneği var, belki biliyorsundur: bir konuyla ilgili tüm kaynakları listeleyen yüksek yıldız sayılı depolar. AI'da kaynak patlaması öyle bir seviyede ki artık awesome listelerinin listeleri türemeye başladı. İşte onlardan biri:

https://github.com/amusi/awesome-ai-awesomeness

Bendeki notlara baktığımda gözüme ilişenler:

- https://github.com/microsoft/AI-For-Beginners (çok iyi görünüyor)

- https://github.com/udlbook/udlbook

- https://github.com/mlabonne/llm-course

- Andrej Karpathy'nin (AI'da önemli bir isim) sunumu: https://www.youtube.com/watch?v=zjkBMFhNj_g&ab_channel=AndrejKarpathy

Teorisi çok yoğun bir alan. Minimum teoriden sonra somut uygulamalara yoğunlaşan öğretici uygulamalar bulmak lazım. Yukarıda verdiğim kaynakları bu açıdan incelemeye çalış. Teori mutlaka gerekiyor ama yetecek kadar[^1]... Akademik bir amacın yoksa daha fazlası suyu bulandırmaktan başka bir şeye yaramaz gibime geliyor.

Bu tartışmayı açık tutalım. Gözüme iliştikçe eklemeler yaparız. Özellikle denenmiş uygulamalı kaynaklar çok önemli. Örneğin basit bir problemi alıp adım adım AI ile çözmek. Bence en kıymetlileri bu pratik kaynaklar. Sen de deneyerek tespit ettiklerini bu zincire eklersin.

[^1]: Teori demişken mesela bir miktar istatistik temelinin de olması lazım. Böyle alt konular var, dikkatle sınırlandırman gereken.

**Reply by Parmendies** [reply]
İşin teorisi için Harvard'ın yayınladığı şöyle bir eğitimi bitirdim: [Harvard CS50’s Artificial Intelligence with Python – Full University Course](https://www.youtube.com/watch?v=5NgNicANyqM). Dili akıcı ve anlaşılabilir, python örnekleri güncel ve verdiğiniz şu depoyla [AI-For-Beginners](https://github.com/microsoft/AI-For-Beginners) aynı müfredat sırasıyla gidiyor gibi görünüyor bu yüzden şimdilik buradan başlıyorum. Yardımlarınız için teşekkür ederim.

Edit: Linkleri düzelttim (roktas)

**Comment by roktas** [comment]
@emres hocanızın da ekleyebileceği kaynaklar olabilir, pingliyorum...

**Comment by roktas** [comment]
İşin mutfağında (Microsoft) çalışan @humak da yardımcı olabilir belki, bir ping de ona...

**Reply by emres** [reply]
Yukarıda çok güzel kaynaklar var. Bunların üzerine çok bir şey eklemek kolay değil. Bazı kaynaklar daha da geniş dallanıp budaklanıyor, belki aşağıda yazacaklarım o geniş çizge içinde mevcuttur ama "keşfedilebilirlik" yolunda galip sayılır mağlup deyip birkaç şey ekleyeyim:

- En son ticari bir projede İngilizce ve Almanca metinleri işlemek için https://spacy.io diye bir Python kitaplığından faydalanmıştım ve epey memnun kalmıştım, özellikle belgelendirmesi çok iyi idi. Minimal kaynakla gayet pratik ve performanslı iş çıkarmayı sağlayan bir sistem idi, halen aktif olarak geliştiriliyor.
- Alanın önde gelen isimlerinden [Christopher Bishop](https://www.microsoft.com/en-us/research/people/cmbishop/) da yeni bir "Deep Learning" kitabı çıkardı geçenlerde, henüz okumadım ama okuma listeme ekledim: "[Deep Learning : Foundations and Concepts](https://www.bishopbook.com/)".
- Şu anda "generative AI" demeyeni "dövüyorlar" ama piyasada şirketlere baktığımda, Belçika, Hollanda, Almanya, vs. gördüğüm halen "tabular" diye nitelendirdiğimiz ve ilişkisel veri tabanlarında yahut CSV, Excel, vb. ortamlarda depolanan türden geleneksel veriden yola çıkarak tahmin işlerini otomatikleştiren sistemler geliştiriliyor. O noktada "deep learning" ne kadar anlamlı, attığı taş ürküttüğü kuşa değiyor mu, tartışmalı. O yüzden [XGBoost](https://xgboost.readthedocs.io/en/stable/) tarzı güçlü makine öğrenme sistemleri halen önemini koruyor.
- Yine ticari ortamlara baktığımda, önemli bir başka konu görüntü işleme ve **endüstriyel** ortamlarda halen yapılacak çok iş var. Orada "derin öğrenme" elbette çok daha güçlü ve bu konuda geliştirmenin en iyi yollarından biri mümkünse o tür bir ortamda staj. Mesela alanında öncü, 150 yıllık, onbinlerce kişiyi çalıştıran uluslararası bir makine üreticisinde çalışırken yer aldığım bir projede, epey pahalı bir **robot** kolun **_ağır motor parçalarını_** kaldırmak için kullandığı klasik görüntü işleme sistemi yerine derin öğrenme ile bir şey geliştirmeye çalışıyorlardı. O makina parçalarında kire, pasa, sıcaklığa, fabrikadaki diğer durumlara göre epey varyasyon vardı ve sistemi eğitmek için gerekli eğitim setini düzgün olarak etiketlenmiş (karmaşık geometrik durumlar söz konusu idi) şekilde oluşturmak bile başlı başına bir proje idi ([Amazon Mechanical Turk](https://www.mturk.com/) üzerinden epey otomasyon işi yapmamız gerekmişti). 
- Yine piyasadan, ticari projelerden bir örnek: zaman serisi **tahmini** (**time series forecasting**). Enerji, yenilenebilir enerji sektöründen tutun, ilaç şirketlerine dek sürekli tahminlerini güncellemeleri gerekiyor. Bunun için dönüp dolaşıp geldikleri alan da zaman serisi konusu. Bu konudaki en pratik ve güncel kitaplardan biri: "[Forecasting: Principles and Practice (3rd ed)](https://otexts.com/fpp3/)"
- Milyarlarca dolarlık cirosu olan uluslararası bir lojistik şirketi şu anda harıl harıl her gün gemilerle gelen 10.000lerce "container"i optimal şekilde nasıl yerleştiririz ve vinçleri, operatör olarak çalışan kişileri en iyi şekilde bunlara nasıl atarız diye mevcut sistemleri inceliyor ve kendisi de bir şeyler geliştirmeye çalışıyor. Bu konu yapay zekada "[constraint programming](https://en.wikipedia.org/wiki/Constraint_programming)" ve benzer isimlerle geçiyor ve bu konudaki popüler sistemlerden biri "[Timefold](https://docs.timefold.ai)". 
- Aynı şirket, bir yandan da aynı problemi "[Reinforcement Learning](https://en.wikipedia.org/wiki/Reinforcement_learning)" ile nasıl çözeriz diye kafa patlatıyor. Kağıt üzerinde ve teoride güzel görünen bu tür yapay zeka sistemlerini endüstriyel ortamda uygulamaya çalışınca türlü çeşit detay ve değişik problemler gündeme geliyor. Bu konudaki en meşhur kaynak:  "[Reinforcement Learning: An Introduction](http://incompleteideas.net/book/the-book-2nd.html)". Buna ek olarak geçenlerde Kevin Patrick Murphy şunu tavsiye etmiş [şu Twitter zincirinde](https://twitter.com/DarceyNLP/status/1771769830388973712): "[An Invitation to Deep Reinforcement Learning](https://arxiv.org/abs/2312.08365)". Murphy'nin önerisi neden dikkatimi çekti? Çünkü Murphy şu güzel kitapların yazarı: "[Probabilistic Machine Learning: An Introduction](https://probml.github.io/pml-book/book1.html)" ve "[Probabilistic Machine Learning: Advanced Topics](https://probml.github.io/pml-book/book2.html)"
- Son olarak: bir şirkette bu konuda öncü projelerde çalışma imkanı henüz olmayanlar için somut ve gerçekçi problemler üzerinde pratik yapmak için en güzel imkanlardan biri [Kaggle](https://www.kaggle.com/) yarışmaları. Şirketler bu ortama epey para verip "elimde şu tür veri setleri var ve şu tür bir problemi çözmeye çalışıyorum" diyor, Kaggle da yarışmacılara bunu sunup belli bir sürede, kısıt koşullara uyan en iyi sonucu elde eden ekibi ödüllendiriyor. **Güzel yanı**, **_tecrübesiz_** olan biri bile bir ekibe dahil olup çok deneyimli uzmanlardan hızlıca bir şeyler kapabilir bir yandan işin bir ucundan tutarken. Sonrasında birinci olanlardan bazıları neyi neden nasıl yaptıklarını da anlatıyorlar, o da çok güzel bir öğrenme kaynağı.

**Reply by roktas** [reply]
Harika eklemeler :clap: Ellerin dert görmesin!

**Reply by emres** [reply]
Bu arada, biraz konuyu dağıtmış gibi olacağım ama **_somut proje önerisi_** olarak: Yıllar önce [Deniz Yüret](http://www.denizyuret.com/) hocanın [Emacs Lisp](https://en.wikipedia.org/wiki/Emacs_Lisp) ile geliştirdiği "deasciifier" kodunu [Python'a aktarmıştım](https://github.com/emres/turkish-deasciifier) (sevgili @roktas üstadım da sağolsun, her zamanki titizliği ile Debian paketlemesi için her şeyi yapmıştı). 

Acaba 2024 itibariyle **daha doğru** sonuçlar veren bir sistem gerçekleştirilebilir mi? Mesela "[Transformer](https://en.wikipedia.org/wiki/Transformer_(deep_learning_architecture))" mimarisi kullanarak? (Bu konuda guzel bir kitap: "[Natural Language Processing with Transformers, Revised Edition](https://www.oreilly.com/library/view/natural-language-processing/9781098136789/)"). Bu soruya dair detayları, ne demek istediğimi şurada etraflıca izah etmeye çalıştım bir süre önce: "[Diacritics restoration: can we do better by using neural networks and deep learning? Perspectives from a 10-year-old open source project](https://ileriseviye.wordpress.com/2020/10/22/diacritics-restoration-can-we-do-better-using-neural-networks-and-deep-learning-perspectives-from-a-10-year-old-open-source-project/)".

**Reply by humak** [reply]
@emres ve hocam @roktas  halihazirda cok guzel kaynaklar paylasmissiniz. 
Ben biraz daha @microsoft tarafinda, sektorde cloud&ai alaninda nasil gelismeler oluyor, nereden takip edilirler, bu konuda -ozellikle Azure ile- cloudda calismak isteyenler icin upskilling kaynaklari paylasayim, ama once biraz onbilgi :):

Cloud & AI:
AI konusuna giris yaparken temelini ogrenmek -bence :)- onemini hala koruyor. AI/ML/DL kavramlari, Prediction & classification kavramlari ve modelleri, hem kavramsal olarak hem de pratik olarak (python ve ai frameworkleri) ogrenmek ve uygulamak guzel bir baslangic olacaktir. Bir focus alani secilip o konuda daha derin bir temel de kurulabilir, mesela Recommendation Systems (ben bu konuda bitirme projemi yapmistim, @emres hocanin da bahsettigi XGBoost ve baska ML algoritmalarini kiyaslayarak. Yine referans verdigi Kaggle ise datasetler icin mukemmel bir kaynak.).

**Ve diyelim ki bu temelimizi olusturduk, sirada ne var, sektorde, cloud ile nasil kullaniliyor AI?**

Bir cok cloud provider bir cok AI urunu/platformu sunuyor aslinda, SaaS, Paas ve/veya IaaS seklinde. Farkli sektorlerden bir cok enterprise-scale sirket, AI konusunda buyuk cloud providerlarin sagladigi bu servislerden yararlaniyor. Benim bundan sonra referans vereceklerim sadece kendi bilgim dahilinde ve Microsoft/ Azure ozelinde olacaktir. (Baska Cloud providerlarin da check edilmesini oneririm)

Microsoft AI ve Azure AI- Saas servislerde ve Cloud'da yapay zeka kullanimi:

- CoPilot'lar, M365'ten Github ve VSCode'a - ve Copilot Studio?: Bu konu AI gelistiricisi olmaktan cok AI'i copilot olarak kullanmak icin asina olmamiz gereken bi konu.  Ama CoPilot Studio diye bir sey de var ki, yine bir MSFT urunu, PowerPlatform dahilinde low code- no code ai-enriched urunler gelistirilmesini de saglayabiliyor.  Ama tabii dedigim gibi, microsoft disinda boyle urunler cikaran cok daha fazla sirket ve girisim var. 
- Azure AI: Gelistiriciler AI'i cloudda nasil kullanabilir?:
          - Cloud native arthitectures: Buyuk sirketler -cloud native olanlar- kendi sistemleriyle entegre olabilecek cozumler icin cloud providerlarinin ai konusunda sunduklariyla hasir nesir.
          - Compute gucu! : Bir cok model ve training gibi asamalar, ve tabii ki data buyuklugu, guclu compute destegi gerektiriyor, bunlari local makinamizda yapmaktansa, cloudda istedigimiz gucte compute sahibi olabiliriz.
          - Alan spesifik cozumler: LLM'ler (gpt modelleri, bkz: azure openai), MLOps ve LLMOps destegi, Computer Vision, Speech, Vector Databases, Bots, bunlarin hepsi icin spesifik urun destekleriyle ya da platformlariyla gelistime yapabiliyor. Bir kac ornek, [AzureML](https://learn.microsoft.com/en-us/azure/machine-learning/overview-what-is-azure-machine-learning?view=azureml-api-2), [Databricks](https://learn.microsoft.com/en-us/azure/databricks/), [AzureOpenAI](https://learn.microsoft.com/en-us/azure/ai-services/openai/), [Vector Databaseler icin AI Search](https://learn.microsoft.com/en-us/azure/search/),  [Speech](https://learn.microsoft.com/en-us/azure/ai-services/speech-service/), gibi gibi.

Upskilling (Azure AI):
- Yukarida belirttigim konularda alistirmalar yapilabilecek cok kaynak mevcut, microsoft learn'deki tutoriallar takip edilebilir. Daha genel ecosystem hakkinda bilgi sahibi olmak istenirse de bir cok kaynak mevcut. Bunlardan bahsetmek gerekirse:
          - [Latest Updates on MSFT AI](https://www.microsoft.com/en-us/ai/blog/)
          -  [Overall AzureAI servisleri](https://learn.microsoft.com/en-us/azure/ai-services/what-are-ai-services)
          - MLOps ve LLMOps: [MLOps nedir? -azure'dan bagimsiz- ](https://towardsdatascience.com/what-is-mlops-everything-you-must-know-to-get-started-523f2d0b8bd8), [LLMOps ](https://techcommunity.microsoft.com/t5/ai-machine-learning-blog/an-introduction-to-llmops-operationalizing-and-managing-large/ba-p/3910996), [MLOps with Azure DevOps](https://github.com/microsoft/MLOps), [ML DevOps aciklayici ve ornek architecture video kaynagi](https://www.youtube.com/watch?v=nst3UAGpiBA)

PS:
- AI ve Etik?: -ilgilisine- kaynaklardan ziyade, ilgili olanlar bu konuyu da arastirabilir. AI hallucination, responsible ai vs.

**Reply by roktas** [reply]
Harika, teşekkürler!
